docs/content/develop/custom-code.md

```diff
@@ -0,0 +1,242 @@
+  post_update: templates/terraform/post_update/PRODUCT_RESOURCE.go.erb
+
+  pre_delete: templates/terraform/pre_delete/PRODUCT_RESOURCE.go.erb
+  post_delete: templates/terraform/post_delete/PRODUCT_RESOURCE.go.erb
+
+  post_import: templates/terraform/post_import/PRODUCT_RESOURCE.go.erb
+```
+
+CRUD operations can be modified with pre/post hooks. This code will be injected directly into the relevant CRUD method as close as possible to the related API call and will have access to any variables that are present when it runs. `pre_create` and `pre_update` run after any [`encoder`]({{< ref "#encoder" >}}). Some example use cases:
+
+- Use `post_create` to set an update-only field after create finishes.
+- Use `pre_delete` to detach a disk before deleting it.
+- Use `post_import` to parse attributes from the import ID and call `d.Set("field")` so that the resource can be read from the API.
+
+### Custom create error handling
+
+```yaml
+custom_code: !ruby/object:Provider::Terraform::CustomCode 
+  post_create_failure: templates/terraform/post_create_failure/PRODUCT_RESOURCE.go.erb
+```
+
+Use `custom_code.post_create_failure` to inject code that runs if a Create request to the API returns an error.
+
+The post_create_failure code will be wrapped in a function like:
+
+```go
+func resourceProductResourcePostCreateFailure(d *schema.ResourceData, meta interface{}) {
+    // Your code will be injected here.
+}
+```
+
+The parameters the function receives are:
+
+- `d`: Terraform resource data. Use `d.Get("field_name")` to get a field's current value.
+- `meta`: Can be cast to a Config object (which can make API calls) using `meta.(*transport_tpg.Config)`
+
+## Replace entire CRUD methods
+
+```yaml
+custom_code: !ruby/object:Provider::Terraform::CustomCode
+  custom_create: templates/terraform/custom_create/PRODUCT_RESOURCE.go.erb
+  custom_update: templates/terraform/custom_update/PRODUCT_RESOURCE.go.erb
+  custom_delete: templates/terraform/custom_delete/PRODUCT_RESOURCE.go.erb
+  custom_import: templates/terraform/custom_import/PRODUCT_RESOURCE.go.erb
+```
+
+Custom methods replace the entire contents of the Create, Update, Delete, or Import methods. For example:
+
+```go
+func resourceProductResourceImport(d *schema.ResourceData, meta interface{}) ([]*schema.ResourceData, error) {
+    // Your code will be injected here.
+}
+```
+
+Custom methods are similar to handwritten code and should be avoided if possible. If you have to replace two or more methods, the resource should be handwritten instead.
+
+## Add extra fields to a resource
+
+Use `custom_code.extra_schema_entry` to add additional fields to a resource. Do not use `extra_schema_entry` unless there is no other option. The extra fields are injected at the end of the resource's [`Schema` field](https://developer.hashicorp.com/terraform/plugin/sdkv2/schemas/schema-types).  They should be formatted as entries in the map. For example:
```

Comment:
Do we cover `virtual_fields` anywhere yet? Almost all `extra_schema_entry` usecases are covered by those, so we could recommend it. `name_prefix` is probably one of the only cases that needs the finer-grained control, and should (long-term) be moved to an override or flag.

Reply:
No, we don't yet cover `virtual_fields` anywhere. Opened b/297044937 to track.